src/nova_tz/keyblob.rs

```diff
@@ -0,0 +1,120 @@
+use std::array::TryFromSliceError;
+use std::convert::TryFrom;
+use std::convert::TryInto;
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Can't parse key blob: {0}")]
+    BadKeyBlob(String),
+}
+
+pub enum DigestPadAlgo {
+    RsaDigestPaddingNone = 0,
+    RsaPkcs115Sha2_256 = 1,
+    RsaPssSha2_256 = 2,
+}
+
+impl TryFrom<u32> for DigestPadAlgo {
+    type Error = Error;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(DigestPadAlgo::RsaDigestPaddingNone),
+            1 => Ok(DigestPadAlgo::RsaPkcs115Sha2_256),
+            2 => Ok(DigestPadAlgo::RsaPssSha2_256),
+            _ => Err(Error::BadKeyBlob(format!(
+                "Bad digest padding algorithm {}",
+                value
+            ))),
+        }
+    }
+}
+
+pub(crate) struct KeyBlob {
+    pub modulus: Vec<u8>,
+    pub public_exponent: Vec<u8>,
+}
+
+// Note (iegor.s): values are copied from qualcomm qseecom module
+const RSA_KEY_SIZE_MAX: usize = 512 + 16;
+const RSA_IV_LENGTH: usize = 16;
+const RSA_HMAC_LENGTH: usize = 32;
+const RSA_KEY_BLOB_LENGHT: usize = 1656;
+
+fn read_u32_le(buf: &[u8]) -> Result<(u32, &[u8]), TryFromSliceError> {
```

Comment:
I do not believe you need these functions.. use a `Cursor` in the caller and the byteorder crate (which is already used in crypto-rs) for endianness readers on that cursor. That way you don't need these functions and just work with a mutable cursor on the slice of data

Reply:
It will lead to ugly repeatable code. because I have to store the buffer of full size anyway until I read real size of it.

Reply:
I'm not sure that's the case.. Do you have a test with a binary that gets decoded? I can try it to see if I can prove you wrong

Reply:
he is correct about reading exponents (see `read_u8_array_dynamic`), because they are length tagged, but the Len tag is after the buffer. So you do need to read the whole buffer first. What you're seeing here includes my "improvements", which essentially re-implemnet a couple pieces of `nom`. 

I also saw other ways to improve this, but I also knew he's against a huge time crunch.

Reply:
the way to do this with pure std+cursor is read max sized dynamic buffer into a vector, and resize it after reading the following length tag

Reply:
Right.. that's what I thought too.. It'll remove all these functions. I don't think that work is critical though so we can live with the implementation here for a bit while we get through the time crunch

Reply:
I've reimplemented it using Cursor. But I also want to remove all that errors (except for 3 field length checks). Because there is total length check at the beginning it's impossible to get Cursor reading errors, right?